### ISSUE TITLE
Voxelization routine

### ISSUE THREAD

**Issue by Kresa96** [kip-hart/MicroStructPy#44]
Very nice work!
We try to use your package MicroStructPy in our research group to built polycristalline electrolytes for impedance simulation. We usually use numpy arrays in which each grain is represented by several array entries (very similar to voxels, or images in 2D). The grains are distinguishable by different integer values in the matrix. In this way it's easy to modify the structure and use it for several different purposes (as an example for our simulations). 

Example for a 2D structure consisting of 2 grains:
1111112222
1111112222
1111222222
1112222222

I think it would be very nice if you could implement a function which can transform the 2d and 3d polygon mesh into numpy arrays. 

Looking forward to your reply！

**Comment by kip-hart** [comment]
Hi thank you! I'm glad to hear you're using MicroStructPy. I understand what you're asking for, where there is a grid of points and the value assigned to each point is the grain number, then these values are stored in a matrix.

I'm not familiar with voxelization approaches. My first thought is to create a list of points then test if point "i" is within grain "j", but that would be very slow. My second thought is to put bounding boxes around each grain, then only test if point "i" is within grain "j" if the point is within the bounding box. This is better, but probably still slow.

Do you have suggestions for voxelization routines?

**Comment by Kresa96** [comment]
Thanks for the fast anwer!
No I am also not very familiar with those methods. But I saw that a voxelization routine for ellipsoidal grains is already implemented in 'Kanapy', a Python package from 'mrgprasad', which was also made to create microstructures in Python. Maybe all you need is in their documentation. 

In 2D I was able to translate the polymesh.plot in a very 'dirty' way into a numpy array: 

polymesh.plot(facecolors = "white", edgecolor = "k")
plt.axis('off')
buf = io.BytesIO()
plt.savefig(buf, format="png", bbox_inches='tight' , pad_inches = 0, dpi=500)
buf.seek(0)
img_arr = np.frombuffer(buf.getvalue(), dtype=np.uint8)
buf.close()
img = cv2.imdecode(img_arr, 0)

s = [[1,1,1],[1,1,1],[1,1,1]]
labeled_mask, num_labels = ndimage.label(img, structure = s)
labeled_mask = expand_labels(labeled_mask, distance = 20)

But in 3D this dirty procedure will not work. 
If you are able to implement something like this, we will probably use your package for lots of stuff.

**Comment by kip-hart** [comment]
Ok got it, I'll check out Kanapy. I've started a branch to add a RasterMesh class. The pseudo-code for creating the mesh is https://github.com/kip-hart/MicroStructPy/blob/555c61e453c62b03905de178a6db2321f049b925/src/microstructpy/meshing/rastermesh.py#L113-L122

This mesh class would have the same attributes as TriMesh. To create the array of grain numbers, my pseudo-code is https://github.com/kip-hart/MicroStructPy/blob/555c61e453c62b03905de178a6db2321f049b925/src/microstructpy/meshing/rastermesh.py#L199-L201

I'd also add functions to write these meshes out in VTK and Abaqus formats.

I can keep you posted with updates. How urgent is the need?

**Comment by Kresa96** [comment]
Nice!
Yeah I think it is a nice idea to implement the output of these files!
Keep me updated!
The need is very urgent 😁

**Comment by kip-hart** [comment]
I've added details in #45 for how this is implemented. You can get the desired array with a script that first creates/opens a PolyMesh, then calls `RasterMesh.from_polymesh`. The raster mesh created by this function has a method `as_array`. Array indices are 0 in the lower-left corner. The functions work in both 2D and 3D.

Upgrading to v1.5.0 will have these features. Please let me know if you encounter any issues.

**Comment by Kresa96** [comment]
Thank you very much for the fast implementation!!
I got a view problems: 
In 2D the plot() method does not work in my case. If I plot the matrix with plt.imshow() for several values of the argument 'mesh_size' I get anomalies in the matrix:
![image](https://user-images.githubusercontent.com/90309466/132648083-ff58c401-1fd9-45ad-ba2a-667120b6bda0.png)

In 3D voxelization does not work at all in my case:
Input:
![image](https://user-images.githubusercontent.com/90309466/132648364-49b610c9-66a0-42ee-8488-86c69910b7f1.png)
Output:
![image](https://user-images.githubusercontent.com/90309466/132648491-0e3a87f5-cb0c-470c-bfce-ef9265eaef04.png)

**Comment by kip-hart** [comment]
Sorry to hear that. Could you attach the 2D and 3D polymesh.txt files? You can also email them to support@microstructpy.org if that doesn't work. I'll debug with those; it's likely a floating point error.

For the `.plot()` function, this is the default behavior when there are no options. If you assign a color to each seed value, then `.plot(color=seed_colors, index_by='attribute')` should produce the desired result. You can also pass other matplotlib arguments, like `.plot(..., edgecolor='k', lw=0.5)`. All the non-MicroStructPy keyword args are passed through to matplotlib. More details at https://docs.microstructpy.org/en/latest/api/meshing/trimesh.html#microstructpy.meshing.TriMesh.plot

**Comment by Kresa96** [comment]
Yes you are right the implemented 2D plot works for me: 
![image](https://user-images.githubusercontent.com/90309466/132681406-386e0065-ac6b-4543-a234-54ba73174d0d.png)
But it seems like that the array gets turned and the anomalies arrise when as_array() gets applied to the grid.

I will send the files to the mail!
Thanks!

**Comment by kip-hart** [comment]
Since (0,0) is in the lower-left corner, plotting with `plt.imshow(..., origin='lower')` should prevent the turning. You can also use the `extent=(left, right, bottom, top)` option to scale the axes into physical space, rather than matrix coordinates

I'll debug the 2D plaid effect and 3D issues now and update you with results.

**Comment by kip-hart** [comment]
I've cleared up the plaid issue in 2D. It was a floating point error. Working on the 3D case now.

<img width="415" alt="Screen Shot 2021-09-09 at 11 44 52 AM" src="https://user-images.githubusercontent.com/41959581/132718612-17525e95-28bf-42c9-ad80-888fd83c0e1e.png">

**Comment by Kresa96** [comment]
Perfect! Thank you very much!

**Comment by kip-hart** [comment]
3D is now working. I'll update the package shortly to v1.5.1 with the fixes.

<img width="469" alt="Screen Shot 2021-09-09 at 12 58 02 PM" src="https://user-images.githubusercontent.com/41959581/132729897-3e23fdbe-e37e-4ee1-ab5d-13209d4b6803.png">

**Comment by Kresa96** [comment]
Thanks a lot I am looking forward to the new version!

**Comment by kip-hart** [comment]
v1.5.1 has been released. Please let me know if you encounter any issues.

**Comment by Kresa96** [comment]
Sorry but the coloring doesnt work in my case: 
![image](https://user-images.githubusercontent.com/90309466/132753287-66f7dbae-cdb4-4a4b-b927-ad42979b2329.png)
Error code:
ValueError                                Traceback (most recent call last)
<ipython-input-25-15df43374e4f> in <module>
     30         colors_list.append(random_color)
     31 
---> 32 rmesh.plot(index_by='attribute', color=['blue', 'red'])

~\anaconda3\envs\micropy4\lib\site-packages\microstructpy\meshing\trimesh.py in plot(self, index_by, material, loc, **kwargs)
   1344                 for i, x in enumerate(pt):
   1345                     grids[i][tuple(pt_tup)] = x
-> 1346             ax.voxels(*grids, inds >= 0, **plt_kwargs)
   1347 
   1348         # Add legend

~\anaconda3\envs\micropy4\lib\site-packages\mpl_toolkits\mplot3d\axes3d.py in voxels(self, facecolors, edgecolors, shade, lightsource, *args, **kwargs)
   2907                     )
   2908 
-> 2909             poly = art3d.Poly3DCollection(
   2910                 faces, facecolors=facecolor, edgecolors=edgecolor, **kwargs)
   2911             self.add_collection3d(poly)

~\anaconda3\envs\micropy4\lib\site-packages\mpl_toolkits\mplot3d\art3d.py in __init__(self, verts, zsort, *args, **kwargs)
    671         and _edgecolors properties.
    672         """
--> 673         super().__init__(verts, *args, **kwargs)
    674         self.set_zsort(zsort)
    675         self._codes3d = None

~\anaconda3\envs\micropy4\lib\site-packages\matplotlib\collections.py in __init__(self, verts, sizes, closed, **kwargs)
   1121             Forwarded to `.Collection`.
   1122         """
-> 1123         Collection.__init__(self, **kwargs)
   1124         self.set_sizes(sizes)
   1125         self.set_verts(verts, closed)

~\anaconda3\envs\micropy4\lib\site-packages\matplotlib\cbook\deprecation.py in wrapper(*inner_args, **inner_kwargs)
    409                          else deprecation_addendum,
    410                 **kwargs)
--> 411         return func(*inner_args, **inner_kwargs)
    412 
    413     return wrapper

~\anaconda3\envs\micropy4\lib\site-packages\matplotlib\collections.py in __init__(self, edgecolors, facecolors, linewidths, linestyles, capstyle, joinstyle, antialiaseds, offsets, transOffset, norm, cmap, pickradius, hatch, urls, offset_position, zorder, **kwargs)
    211 
    212         self._path_effects = None
--> 213         self.update(kwargs)
    214         self._paths = None
    215 

~\anaconda3\envs\micropy4\lib\site-packages\matplotlib\artist.py in update(self, props)
    996                         raise AttributeError(f"{type(self).__name__!r} object "
    997                                              f"has no property {k!r}")
--> 998                     ret.append(func(v))
    999         if ret:
   1000             self.pchanged()

~\anaconda3\envs\micropy4\lib\site-packages\matplotlib\collections.py in set_color(self, c)
    749             For setting the edge or face color individually.
    750         """
--> 751         self.set_facecolor(c)
    752         self.set_edgecolor(c)
    753 

~\anaconda3\envs\micropy4\lib\site-packages\mpl_toolkits\mplot3d\art3d.py in set_facecolor(self, colors)
    792 
    793     def set_facecolor(self, colors):
--> 794         PolyCollection.set_facecolor(self, colors)
    795         self._facecolor3d = PolyCollection.get_facecolor(self)
    796 

~\anaconda3\envs\micropy4\lib\site-packages\matplotlib\collections.py in set_facecolor(self, c)
    778         """
    779         self._original_facecolor = c
--> 780         self._set_facecolor(c)
    781 
    782     def get_facecolor(self):

~\anaconda3\envs\micropy4\lib\site-packages\matplotlib\collections.py in _set_facecolor(self, c)
    762         except AttributeError:
    763             pass
--> 764         self._facecolors = mcolors.to_rgba_array(c, self._alpha)
    765         self.stale = True
    766 

~\anaconda3\envs\micropy4\lib\site-packages\matplotlib\colors.py in to_rgba_array(c, alpha)
    339         return np.zeros((0, 4), float)
    340     else:
--> 341         return np.array([to_rgba(cc, alpha) for cc in c])
    342 
    343 

~\anaconda3\envs\micropy4\lib\site-packages\matplotlib\colors.py in <listcomp>(.0)
    339         return np.zeros((0, 4), float)
    340     else:
--> 341         return np.array([to_rgba(cc, alpha) for cc in c])
    342 
    343 

~\anaconda3\envs\micropy4\lib\site-packages\matplotlib\colors.py in to_rgba(c, alpha)
    187         rgba = None
    188     if rgba is None:  # Suppress exception chaining of cache lookup failure.
--> 189         rgba = _to_rgba_no_colorcycle(c, alpha)
    190         try:
    191             _colors_full_map.cache[c, alpha] = rgba

~\anaconda3\envs\micropy4\lib\site-packages\matplotlib\colors.py in _to_rgba_no_colorcycle(c, alpha)
    263         raise ValueError(f"Invalid RGBA argument: {orig_c!r}")
    264     if len(c) not in [3, 4]:
--> 265         raise ValueError("RGBA sequence should have length 3 or 4")
    266     if not all(isinstance(x, Number) for x in c):
    267         # Checks that don't work: `map(float, ...)`, `np.array(..., float)` and

ValueError: RGBA sequence should have length 3 or 4

**Comment by kip-hart** [comment]
This will be fixed in v1.5.2. For 3D, you'll need to use the keyword `facecolors=` instead of `colors=`.

Specifying `facecolors=['blue', 'red']` only gives colors to seeds 0 and 1. In this case, other seeds will be rendered invisible:

<img width="466" alt="Screen Shot 2021-09-09 at 4 27 13 PM" src="https://user-images.githubusercontent.com/41959581/132757812-92ec3921-86e6-454c-9016-a343a4daf975.png">

What I've done in the past is create a color list like:

```python
n = max(rmesh.element_attributes) + 1
colors = ['C' + str(i % 10) for i in range(n)]
rmesh.plot(index_by='attribute', facecolors=colors)
plt.show()
```

This will produce a plot like:

<img width="456" alt="Screen Shot 2021-09-09 at 4 31 19 PM" src="https://user-images.githubusercontent.com/41959581/132758359-618da805-c74c-4742-81a9-f513f37af68a.png">

You can also create a color list by sampling colormaps. Additionally, you can set grains of the same material phase to have the same color.

Hopefully this helps. Version 1.5.2 will be available shortly.

**Comment by Kresa96** [comment]
It works!  Great work and very fast implementation! 
Thank you very much!
